Проверим, что у вас подключена возможность работы с SQL

In [2]:
%load_ext sql
%sql sqlite://

Создадим таблицу специальностей со следующими полями:
1. id - уникальный идентификатор специальности
2. Name - Специальность

In [3]:
%%sql
DROP TABLE IF EXISTS Specialities;

CREATE TABLE Specialities
(
    id integer PRIMARY KEY,
    name varchar(50)
);

 * sqlite://
Done.
Done.
Done.


[]

Проверим, что таблица создана

In [4]:
%sql select * from Specialities

 * sqlite://
Done.


[]

Заполним таблицу данными

In [5]:
%%sql
INSERT INTO Specialities(name) VALUES -- Автоинкрементация по PRIMARY KEY
('Биотехнические системы и технологии'),
('Компьютерная безопасность'),
('Прикладная математика и информатика'),
('Мехатроника и робототехника'),
('Системный анализ и управление'),
('Информационная безопасность телекоммуникационных систем'),
('Управление в технических системах');

 * sqlite://
7 rows affected.


[]

SQLite требует включение поддержки внешних ключей.
Более подробно можно посмотреть [здесь](https://sqlite.org/foreignkeys.html)

In [6]:
%%sql
PRAGMA foreign_keys = ON;

 * sqlite://
Done.


[]

Создадим таблицу Студентов со следующими полями:
1. Student_id - уникальный идентификатор студента
2. Name - имя студента
3. Second Name - фамилия студента
4. Speciality - специальность студента
5. StudyGroup - группа студента
6. GPA - средний балл

In [20]:
%%sql
DROP TABLE IF EXISTS Students;

CREATE TABLE Students
(
    student_id integer PRIMARY KEY,
    name varchar(50),
    secondName varchar(50),
    speciality_id integer,
    studyGroup varchar(20),
    GPA decimal(1,1),
    FOREIGN KEY (Speciality_id) REFERENCES Specialities(id)
);

 * sqlite://
Done.
Done.


[]

Проверим, что таблица создана

In [9]:
%sql select * from Students

 * sqlite://
Done.


[]

Заполним таблицу данными

In [21]:
%%sql
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (1, 'Альянов', 'Николай', 1, '04-25', 3.58);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (2, 'Киселева', 'Ева', 1, '01-25', 4.58);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (3, 'Исаева', 'Анна', 1, '01-25', 4.75);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (4, 'Егоров', 'Павел', 2, '01-25', 4.83);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (5, 'Тихомиров', 'Тимофей', 2, '03-25', 4.18);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (6, 'Ширшов', 'Александр', 2, '06-25', 3.91);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (7, 'Фатеев', 'Максим', 3, '05-25', 3.8);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (8, 'Киселев', 'Петр',  3, '05-25', 3.5);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (9, 'Агашкина', 'Дарья', 4, '02-25', 4.18);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (11, 'Комарова', 'Екатерина', 6, '02-25', 4.83);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (12, 'Челыхов', 'Иван', 6, '06-25', 4.45);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (13, 'Назаров', 'Александр', 6, '06-25', 3.73);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (14, 'Горлова', 'Ксения', 7, '01-25', 5);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (17, 'Горин', 'Олег', 5, '01-25', 3.89);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (18, 'Остапенко', 'Алексей', 5, '04-25', 3.81);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (19, 'Антонникова', 'Валерия', 5, '04-25', 4.89);
insert into Students(student_id, SecondName, Name, Speciality_id, StudyGroup, GPA) VALUES (20, 'Котова', 'Евгения', 3, '05-25', 4.29);

 * sqlite://
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.


[]

Проверим наличие данных

In [22]:
%sql select * from Students;

 * sqlite://
Done.


[(1, 'Николай', 'Альянов', 1, '04-25', 3.58),
 (2, 'Ева', 'Киселева', 1, '01-25', 4.58),
 (3, 'Анна', 'Исаева', 1, '01-25', 4.75),
 (4, 'Павел', 'Егоров', 2, '01-25', 4.83),
 (5, 'Тимофей', 'Тихомиров', 2, '03-25', 4.18),
 (6, 'Александр', 'Ширшов', 2, '06-25', 3.91),
 (7, 'Максим', 'Фатеев', 3, '05-25', 3.8),
 (8, 'Петр', 'Киселев', 3, '05-25', 3.5),
 (9, 'Дарья', 'Агашкина', 4, '02-25', 4.18),
 (11, 'Екатерина', 'Комарова', 6, '02-25', 4.83),
 (12, 'Иван', 'Челыхов', 6, '06-25', 4.45),
 (13, 'Александр', 'Назаров', 6, '06-25', 3.73),
 (14, 'Ксения', 'Горлова', 7, '01-25', 5),
 (17, 'Олег', 'Горин', 5, '01-25', 3.89),
 (18, 'Алексей', 'Остапенко', 5, '04-25', 3.81),
 (19, 'Валерия', 'Антонникова', 5, '04-25', 4.89),
 (20, 'Евгения', 'Котова', 3, '05-25', 4.29)]

In [23]:
%sql select * from Specialities;

 * sqlite://
Done.


[(1, 'Биотехнические системы и технологии'),
 (2, 'Компьютерная безопасность'),
 (3, 'Прикладная математика и информатика'),
 (4, 'Мехатроника и робототехника'),
 (5, 'Системный анализ и управление'),
 (6, 'Информационная безопасность телекоммуникационных систем'),
 (7, 'Управление в технических системах')]

## Задание 1. Напишите SQL-запрос, возвращающий фамилии всех студентов с именем Александр

## Задание 2. Выведите все названия специальностей, на которых учится хотя бы один студент со средним баллом 4.5 и выше. Имя специальности должно встречаться один раз в выводе

## Задание 3. Вывести только тех студентов, фамилия которых начинается с буквы 'к', а в имени есть буква 'е'.

## Задание 4. Вывести специальности (+ их средние баллы), количество групп в которых больше одной

## Задание 5. Для всех специальностей вывести средний балл по специальности для студентов, в чьей фамилия вторая буква НЕ 'а', при условии, что количество таких студентов на специальности не менее 2. В обратном случае, специальность выводить не требуется